### PR TITLE
Log via glaze instead of JSON strings, and templates

### PR DESCRIPTION
**Pull request by andrlime**

### CODE REVIEW

**Comment by andrlime** [comment]
I'm a little sus of the templates I wrote

File: exchange/src/utils/logger/logger.cpp [hunk 1]
```diff
@@ -17,35 +18,50 @@ Logger::get_logger()
...
+    auto nowAsTm = *std::localtime(&nowAsTimeT);
+
+    auto ms = duration_cast<milliseconds>(now.time_since_epoch()) % 1000;
+
+    // Use stringstream to format the time
+    std::ostringstream timestamp;
+    timestamp << std::put_time(&nowAsTm, "%Y-%m-%d %H:%M:%S");
+    timestamp << '.' << std::setfill('0') << std::setw(3) << ms.count();
+
+    return timestamp.str();
```
Comment (egelja) [inline exchange/src/utils/logger/logger.cpp:39]:
```suggestion
    using namespace std::chrono;

    // Get time from Quill
    auto now = quill::Clock::now();

    // Get epoch time in milliseconds, and create time point
    auto epoch_millis = duration_cast<milliseconds>(now.time_since_epoch());
    auto system_now = time_point<system_clock>{epoch_millis};

    return fmt::format("{:%Y-%m-%d %H:%M:%S}", system_now).substr(0, 23);
```

https://godbolt.org/z/oK46fK1xa

File: exchange/src/utils/logger/logger.hpp [hunk 2]
```diff
@@ -11,6 +11,26 @@
 namespace nutc {
 namespace events {
 
+std::string timestamp_in_ms();
+
+template <typename T>
+concept MetaSpecialized = requires {
```
Comment (egelja) [inline exchange/src/utils/logger/logger.hpp:17]:
```suggestion
concept GlazeMetaSpecialized = requires {
```

File: exchange/src/utils/logger/logger.hpp [hunk 3]
```diff
@@ -11,6 +11,26 @@
...
+    {
+        glz::meta<T>::value
+    } -> std::convertible_to<decltype(glz::meta<T>::value)>;
+};
+
+template <MetaSpecialized T>
+struct WithTimestamp {
+    T data;
```
Comment (egelja) [inline exchange/src/utils/logger/logger.hpp:25]:
```suggestion
    const T& data;
```

don't take a copy

File: exchange/src/utils/logger/logger.hpp [hunk 4]
```diff
@@ -11,6 +11,26 @@
...
+    std::string timestamp;
+
+    // Constructor to initialize the base data and timestamp
+    WithTimestamp(const T& data) :
```
Comment (egelja) [inline exchange/src/utils/logger/logger.hpp:29]:
```suggestion
    explicit WithTimestamp(const T& data) :
```

**Review by egelja — CHANGES_REQUESTED** [review]

**Review by egelja — COMMENTED** [review]
does this run

**Comment by andrlime** [comment]
it builds yeah

**Comment by egelja** [comment]
On 11/17/2023 14:14, Andrew Li wrote:
> it builds yeah

run != build

-- 
Nikola Maruszewski
Don't Panic - The Hitchhiker's Guide to the Galaxy.
Northwestern University | Class of 2025
BS in Computer Science | MS in Computer Engineering
***@***.***
***@***.***
https://nikola.cx

File: exchange/src/utils/logger/logger.cpp [hunk 5]
```diff
@@ -17,35 +19,47 @@ Logger::get_logger()
...
-    }
+    // Log to the main log too
+    log_i(events, "Logging event {}", buffer);
 
-    output_file_ << " }\n"; // close the brace and end the line
+    // Write start of JSON
+    output_file_ << buffer << "\n";
 }
 
+// Explicit instantiations
```
Comment (egelja) [inline exchange/src/utils/logger/logger.cpp:59]: Why explicit?

**Review by egelja — CHANGES_REQUESTED** [review]

File: exchange/src/utils/logger/logger.cpp (hunk 5 above)
Comment (andrlime) [reply exchange/src/utils/logger/logger.cpp:59]: steve did this, I have no idea

Comment (stevenewald) [reply exchange/src/utils/logger/logger.cpp:59]: Because we're separating header and implementation of a template function

Comment (egelja) [reply exchange/src/utils/logger/logger.cpp:59]: Again

W H Y

Comment (stevenewald) [reply exchange/src/utils/logger/logger.cpp:59]: Because we know the types that the template will be instantiated for so why not?

Comment (egelja) [reply exchange/src/utils/logger/logger.cpp:59]: Fine

**Comment by stevenewald** [comment]
hit wrong button lol

**Review by egelja — APPROVED** [review]
lgtm

**Pull request merged**